### DISCUSSION TITLE
How to set/use $title$ in `format.pdf.include-before-body` includes?

### DISCUSSION THREAD

**Discussion by putzwasser** [quarto-dev/quarto-cli#10799] [Q&A]
### Description

Until now, I only worked with a quarto book project. Now I've set up a “self-contained” file to render. My issue is that I can't access metadata field like `title` in includes in this set-up.

Why isn't `title` populated in this set-up / How would I use metadata correctly, here?

```md
---
author: John Doe
title: "My Fancy Document"
format:
  pdf:
    documentclass: scrartcl
    include-in-header:
      - text: \usepackage{blindtext,scrlayer-scrpage}\clearpairofpagestyles\ohead{$title$}\ihead{Lorem Ipsum}
---
\blinddocument
```

![image](https://github.com/user-attachments/assets/2bb65553-4a0b-4dc8-b30b-8293e2085170)

I expected `My Fancy Document` to show up in the header. What am I missing?

```
$ quarto --version
1.5.57
$ uname -r
6.6.47-1-MANJARO
```

**Comment by mcanouil** [comment]
You can't. You need to use partials as documented, see <https://quarto.org/docs/journals/templates.html#overview>.

Also, do not use raw language without telling Quarto/Pandoc you are using it, see <https://quarto.org/docs/authoring/markdown-basics.html#raw-content>.

**Comment by putzwasser** [comment]
> You need to use partials as documented

I wasn't aware of that. Thanks.

> do not use raw language without telling Quarto/Pandoc you are using it

I'm not entirely sure what mean, even after reading the section you linked. And why is that relevant? Quarto seems to parse that just fine 🤔 I'm confused 😅

**Reply by mcanouil** [reply]
> I wasn't aware of that. Thanks.

Then where did you learn about `$title$`? That's Pandoc template/partials system.

> Quarto seems to parse that just fine 🤔 I'm confused 😅

Until Quarto/Pandoc doesn't. You are free to take your chances for future headaches.🤷‍♂️
If raw code block exist, it's for a reason.

**Reply by gordonwoodhull** [reply]
> If raw code block exist, it's for a reason.
 
I believe currently you must _not_ use raw blocks inside `include-in-header`. 

Or to put it differently, the content of `include-in-header` is always raw text of the output language, and attempting to use a raw block there will cause the raw block syntax to be output verbatim.

**Reply by mcanouil** [reply]
Indeed, I was talking about the document content not the content provided through metadata.

**Comment by Steinthal** [comment]
While I agree to use partials, especially if you want to reuse a format, you can use the "\newcommand" trick mentioned in this forum some time ago (#5526, #5538):

````
---
author: John Doe
title: "My Fancy Document"
format: 
  pdf:
    documentclass: scrartcl
    include-in-header:
      - text: |
          \newcommand{\mytitle}{{}{{< meta title >}}{}}
          \usepackage{scrlayer-scrpage}\clearpairofpagestyles\ohead{\mytitle}\ihead{Lorem Ipsum}
---

{{< lipsum 1 >}}

{{< pagebreak >}}

{{< lipsum 1 >}}
````

BTW: Since Quarto as of 1.5.57 provides the "[Lorem Ipsum Text Shortcodes](https://quarto.org/docs/authoring/lipsum.html)", you don't need the "blindtext" package.

**Reply by mcanouil** [reply]
@Steinthal even if what you shared works it does not answer the question which was about using `$title$` (not `title`).
There are quite few other approaches here.

**Reply by Steinthal** [reply]
@mcanouil , I think the OP's problem is "I can't access metadata field like title". He writes he "expected `My Fancy Document` to show up in the header." -- like this: 

![CleanShot 2024-09-15 at 19 32 25@2x](https://github.com/user-attachments/assets/268ab1bc-1f3a-4662-940e-26fb0725dcf6)

I think "$title$" was just his clunky attempt to add it. His question above the code says "Why isn't *title* populated in this set-up / How would I use *metadata* correctly, here?" (emphasis added by me).

**Reply by mcanouil** [reply]
I know, but still the title and code example are what they are.
As I said, there are many ways (not limited to the two here) to achieve that.
As the use case was not explained/stated, you can make many assumptions but `$keyword$` is a partial/template thing which is the best way to make advanced modifications.
`include-` are designed for small additions to the default template, not really to override it as it could lead to unexpected results to have several times a command used.
So, for the particular request here, I stand by the fact that using partials is the safest way to go rather than using `include-` to add extra commands using a shortcode in it.

**Reply by Steinthal** [reply]
As I wrote, I agree to use partials and have used them to prepared my own template. However, the \newcommand trick is handy to populate the (running) heading for PDF files. For example, if you have to deal with different information from meta data in different constellations (to automate something). It would be unwieldy to prepare an own template for each case.

**Reply by mcanouil** [reply]
You don't need to define a complete template, that is the whole point of partials: https://quarto.org/docs/journals/templates.html#latex-partials

That was the following extension is leveraging to make the title page: https://github.com/nmfs-opensci/quarto_titlepages/tree/main/_extensions/titlepage

**Comment by nithinmkp** [comment]
i am exactly having a problem here. in which part of the partial should running header be defined?

**Reply by mcanouil** [reply]
Did you look at the partials?

**Reply by nithinmkp** [reply]
yes. I did modify author template using partials. but could not really figure out running header part. would be kind of you to show an example

**Reply by mcanouil** [reply]
I don't know what you want to do and since position/order matters in LaTeX, I can't help you.
As I said, take a look at the source code of the partials: https://quarto.org/docs/journals/templates.html#latex-partials

There are nothing more to it than that. The files are 95% LaTeX code. The rest is Pandoc variable templating.
If you want to modify LaTeX template/partials, it means you know LaTeX, if not, Quarto repository is not the place to learn unfortunately.

**Reply by Steinthal** [reply]
> yes. I did modify author template using partials. but could not really figure out running header part. would be kind of you to show an example

You are not very specific about the running header. Maybe this will help:

Create an include "_myheader.tex" that you add in "pandoc.tex" (e.g. before "$for(header-includes)$".
In "_myheader.tex" add something like this* which gives you some flexibility (adjust as necessary):

````latex
\usepackage[headsepline,footsepline]{scrlayer-scrpage}

\newpairofpagestyles{my-header-footer}{
\clearpairofpagestyles
\ihead*{$if(header-left)$$header-left$$else$$title$$endif$}
\chead*{$if(header-center)$$header-center$$else$$endif$}
\ohead*{$if(header-right)$$header-right$$else$$date$$endif$}
\ifoot*{$if(footer-left)$$footer-left$$else$$for(author)$$author$$sep$, $endfor$$endif$}
\cfoot*{$if(footer-center)$$footer-center$$else$$endif$}
\ofoot*{$if(footer-right)$$footer-right$$else$\thepage$endif$}
}
\pagestyle{my-header-footer}
\automark{section}
````

In your qmd-file define e.g.:

````yaml
header-left: \leftmark
````

* That's basically how header and footer are defined in the [Eisvogel template](https://github.com/Wandmalfarbe/pandoc-latex-template)

**Comment by nithinmkp** [comment]
I want to define a custom short title to YAML, pass this as running header in my article

````={qmd}
---
title: test
authors:
  - name: Norah Jones
    affiliation: The University
    roles: writing
    corresponding: true
bibliography: references.bib
short-title: "short title"
---

## Section
This is a simple placeholder for the manuscript's main document [@knuth84].
````
the _quarto.yml file

````={yml}
project:
  type: manuscript

manuscript:
  article: index.qmd

format:
  pdf:
    documentclass: scrartcl
    include-in-header: 
         - file: "preamble.tex"

  # (other formats)
  # pdf: default

execute:
  freeze: true
````
 
also my latex preamble

````={tex}
\usepackage[automark, headsepline]{scrlayer-scrpage}
\clearpairofpagestyles
\ohead*{$short-title$}

\cfoot*{\pagemark}
````

**Reply by Steinthal** [reply]
Your example does work for me. Which problem do you have?

**Reply by nithinmkp** [reply]
The issue is it is not interpolating the string I pass as short title. It just passes short title no matter what the string is

**Reply by mcanouil** [reply]
Because you used include.
Variables only work in partials/templates.

**Comment by nithinmkp** [comment]
Thanks!! That Worked
